geoapps/interpolation/driver.py

```diff
@@ -0,0 +1,207 @@
+            ):
+                if isinstance(entity, ObjectBase):
+                    topo_obj = entity
+
+            if getattr(topo_obj, "vertices", None) is not None:
+                topo = topo_obj.vertices
+            else:
+                topo = topo_obj.centroids
+
+            if self.params.topography["data"] is not None:
+                topo[:, 2] = self.params.geoh5.get_entity(
+                    self.params.topography["data"].uid
+                )[0].values
+
+            lin_interp = LinearNDInterpolator(topo[:, :2], topo[:, 2])
+            z_interp = lin_interp(xyz_out_orig[:, :2])
+
+            ind_nan = np.isnan(z_interp)
+            if any(ind_nan):
+                tree = cKDTree(topo[:, :2])
+                _, ind = tree.query(xyz_out_orig[ind_nan, :2])
+                z_interp[ind_nan] = topo[ind, 2]
+
+            top = xyz_out_orig[:, 2] > z_interp
+            if self.params.max_depth is not None:
+                bottom = np.abs(xyz_out_orig[:, 2] - z_interp) > self.params.max_depth
+        elif self.params.max_depth is not None:
+            bottom = xyz_out_orig[:, 2] > self.params.max_depth
+
+        for key in values_interp.keys():
+            values_interp[key][top] = self.params.no_data_value
+            values_interp[key][bottom] = self.params.no_data_value
+
+        if self.params.xy_extent is not None and self.params.geoh5.get_entity(
+            self.params.xy_extent
+        ):
+            xy_ref = self.params.xy_extent
+            if hasattr(xy_ref, "centroids"):
+                xy_ref = xy_ref.centroids
+            elif hasattr(xy_ref, "vertices"):
+                xy_ref = xy_ref.vertices
+
+            tree = cKDTree(xy_ref[:, :2])
+            rad, _ = tree.query(xyz_out_orig[:, :2])
+            for key in values_interp.keys():
+                if self.params.max_distance is not None:
+                    values_interp[key][
+                        rad > self.params.max_distance
+                    ] = self.params.no_data_value
+
+        for key in values_interp.keys():
+            if dtype[key] == np.dtype("int32"):
+                primitive = "integer"
+                vals = np.round(values_interp[key]).astype(dtype[key])
+            else:
+                primitive = "float"
+                vals = values_interp[key].astype(dtype[key])
+
+            self.params.out_object.add_data(
+                {key + self.params.ga_group_name: {"values": vals, "type": primitive}}
+            )
+
+        if self.params.monitoring_directory is not None and path.exists(
+            self.params.monitoring_directory
+        ):
+            monitored_directory_copy(
+                self.params.monitoring_directory, self.params.out_object
+            )
+
+
+if __name__ == "__main__":
+    print("Loading geoh5 file . . .")
+    file = sys.argv[1]
+    ifile = InputFile.read_ui_json(file)
+    params = DataInterpolationParams(ifile)
+    params.geoh5.close()
```

Comment:
haha, once 556 get merged to devlop I think we can remove this again since the read_ui_json method closes the workspace